<!DOCTYPE html>
<html>
<head>
<!-- The title is shown in the browser tab -->
  <title>Inner_Outer_Glomeruli</title>

<!-- Some required metadata and viewport which is better than pixel as it adjust to the resolution -->
<!-- 100vw (viewport) is the max. resolution, e.g. at 800px (pixel) screen its 800 -->
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="viewport" content="height=device-height, initial-scale=1.0">
  <link href="foamtree-3.5.0-demo/demos/assets/css/bootstrap.min.css" rel="stylesheet" />
  <link href="foamtree-3.5.0-demo/demos/assets/css/common.css" rel="stylesheet" />


<style>
/* Styles for font size (p = standard, h1 are headings), color, ... */
/* h1="Template" */
/* h2="How to read the treemap", "Gene name in Details" */
/* h3="Loading requires between 20 - 60 seconds" */
/* h4="Normalized Counts" in Details graph */

  body {background-color:rgb(100,100,100);}
  h1 {color: white; font-size: 1.9vw}
  h2 {color: white; font-size: 1.1vw}
  h3 {color: white; font-size: 4vw}
  h4 {color: rgb(155,155,155); font-size: 0.6vw;}
  p {color: white; font-size: 0.8vw}


/* Here we "draw" the boxes with size, position, color, ... */
/* Styles for the foamtree-box ("#left") and the legend-box ("#right") */
/* width in viewport (vw); 100vw = browser width in pixel; therefore all elements in vw adjust with browser size */

#left {
  position: absolute;
  width: 67.5vw;
  height: 45vw;
  left: 0vw;
  background-color:rgb(60,60,60);
  z-index: +2;
  display: block;
}

#right {
  position: absolute;
  width: 29vw;
  height: 45vw;
  left: 68.5vw;
  background-color: rgb(100,100,100);
  display: block;
}


/* Styles for the details panel */
/* "position absolute" to stay on its fixed position; "top" & "left" define this position */

#details {
  position: absolute;
  width: 30vw;
  height: 38vw;
  top: 7vw;
  left: 37.5vw;
  text-align: left;
  background-color: rgb(60,60,60);
  z-index: +1;
  display: block;
  overflow-y: scroll;
  padding-left: 1%;
  padding-right: 1%;
}


/* Styles for the Search-button; similar to btn-default but with viewport (vw) instead of pixel (px) */

.Search-button {
  height: 3vw
  display:inline-block;
  margin-bottom:0;
  font-weight:400;
  text-align:center;
  white-space:nowrap;
  vertical-align:middle;
  -ms-touch-action:manipulation;
  touch-action:manipulation;
  cursor:pointer;
  background-image:none;
  border:0vw solid transparent;
  font-size:1vw;
  border-radius:0.1vw;
  -webkit-user-select:none;
  -moz-user-select:none;
  -ms-user-select:none;
  user-select:none;
  color:#333;
  background-color:#fff;
  border-color:#ccc
}


/* Styles for the Search bar and containing box */

.input-group {
  position: absolute;
  width: 20vw;
}

.Input-container {
  width: 29vw;
  height: 3vw;
}


/* Styles for the Search autocomplete input */

      .twitter-typeahead {
        display: block !important;
      }
      .tt-hint {
      	display: block;
      	line-height: 1.428571429;
      	color: #999;
      	vertical-align: middle;
      	background-color: #ffffff;
      }
      .tt-dropdown-menu {
        width: 100%;
      	min-width: 20vw;
      	margin-top: 0.1vw;
      	padding: 0.14vw 0;
      	background-color: #ffffff;
      	border: 0.05vw solid #cccccc;
      	border: 0.05vw solid rgba(0, 0, 0, 0.15);
      	border-radius: 0.2vw;
      	background-clip: padding-box;
        max-height: 37vw;
        overflow-y: auto;
      }
      .tt-suggestion {
      	display: block;
      	padding: 0.16vw 1vw;
      }
      .tt-suggestion > div > small {
        font-size: 100%;
        color: #888;
        margin-left: 0.31vw;
        padding-left: 0.31vw;
        border-left: 0.05vw solid #eee;
      }
      .tt-suggestion:hover {
        background-color: #f5f5f5;
      }


/* Styles for loading screen and loading icon with animation ("spinner") */

.loading {
  position: absolute;
  left: 9vw;
  top: 10vw;
  text-align: center;
  z-index: +3;
}

.spinner {
  margin: 5vw auto;
  width: 5vw;
  height: 4vw;
  text-align: center;
  font-size: 1vw;
  z-index: +3;
}

.spinner > div {
  background-color:rgb(256,256,256);
  height: 100%;
  width: 0.6vw;
  display: inline-block;
  
  -webkit-animation: sk-stretchdelay 1.2s infinite ease-in-out;
  animation: sk-stretchdelay 1.2s infinite ease-in-out;
}

.spinner .rect2 {
  -webkit-animation-delay: -1.1s;
  animation-delay: -1.1s;
}

.spinner .rect3 {
  -webkit-animation-delay: -1.0s;
  animation-delay: -1.0s;
}

.spinner .rect4 {
  -webkit-animation-delay: -0.9s;
  animation-delay: -0.9s;
}

.spinner .rect5 {
  -webkit-animation-delay: -0.8s;
  animation-delay: -0.8s;
}

@-webkit-keyframes sk-stretchdelay {
  0%, 40%, 100% { -webkit-transform: scaleY(0.4) }  
  20% { -webkit-transform: scaleY(1.0) }
}

@keyframes sk-stretchdelay {
  0%, 40%, 100% { 
    transform: scaleY(0.4);
    -webkit-transform: scaleY(0.4);
  }  20% { 
    transform: scaleY(1.0);
    -webkit-transform: scaleY(1.0);
  }
}

</style>


<!-- Defines the dataset for the details panel -->

<div id="datasets"></div>

</head>


<body>
<!-- First we load some required Scripts and then we fill the boxes created in Style -->
<!-- Script for BarPlot (Chart.js)-->
  <script src="data/chart_v2.8.0.js"></script>
<!-- Script for FoamTree visualization -->
  <script src="foamtree-3.5.0-demo/carrotsearch.foamtree.js"></script>
<!-- Scripts for tree model utilities -->
  <script src="foamtree-3.5.0-demo/carrotsearch.foamtree.util.treemodel.js"></script>
  <script src="foamtree-3.5.0-demo/demos/assets/js/carrotsearch.jsonp.js"></script>
  <script src="foamtree-3.5.0-demo/demos/assets/js/carrotsearch.template.js"></script>
  <script src="foamtree-3.5.0-demo/demos/assets/js/jquery-2.0.3.min.js"></script>
  <script src="foamtree-3.5.0-demo/demos/assets/js/typeahead.js"></script>
<!-- Include Hammer.js for mobile interactions. Not required for desktop-only apps. -->
  <script src="foamtree-3.5.0-demo/demos/assets/js/hammer.min.js"></script>


<!-- Loading screen with text (heading3 = <h3>) and line break <br> -->

<div class="loading">
  <h3>Loading requires between<br>20 - 60 seconds</h3>

  <div class="spinner">
    <div class="rect1"></div>
    <div class="rect2"></div>
    <div class="rect3"></div>
    <div class="rect4"></div>
    <div class="rect5"></div>
  </div>
</div>


<!-- Box which will be filled with the foamtree vizualisation after loading -->

<div id="left">
</div>




<!-- Legend-box with heading, Search input-bar & buttons -->

<div id="right">
  <h1>Glomeruli -   
    <span style="color:rgb(30, 144, 255)"><b>Inner</b></span>
    vs
    <span style="color:rgb(238, 44, 44)"><b>Outer</b></span>
  </h1>

  <div class="Input-container"/> 
    <form id="form" class="form-group">
      <div class="input-group">
        <input name="query" id="search" class="form-control" style="width: 20vw; height: 2vw;" placeholder="Search for genes"/>
        <span class="input-group-btn">
          <button class="Search-button" style="height: 2vw;" type="submit">Go!</button>
        </span>
      </div>
    </form>

    <button id="detailsbutton" class="Search-button" style="position: absolute;height: 2vw;left: 23vw" onclick="showdetails()" value="A">Details >>></button>
  </div>


<!-- Legend text with hyperlink (href=) and image (path: data/Legend.png)  -->

  <p>Glomeruli were isolated from extremely preterm baboons (Papio anubis) that received two weeks of neonatal intensive care, including 
     mechanical ventilation, similar to human protocols. The baboons were delivered via elective cesarean section at 125 days of gestation, 
     equivalent to 24 weeks in humans, the earliest point of viability.
     Laser microdissection of frozen tissue slides was used to separate 
     <span style="color:rgb(30, 144, 255)"><b>inner, prenatal</b></span>
     and
     <span style="color:rgb(238, 44, 44)"><b>outer, postnatal</b></span>
     <b>glomeruli</b>.
  </p>

  <h2><b>How to read the treemap</b></h2>

  <p>Each transcript / transcript cluster is represented by an individual polygon / tile.<br>
     The mean normalized transcript count is represented by the area of a tile, the <br>
     <i>r</i>-fold change by its color. Geneset size is calculated as the sum of the values of every transcript in this group;  
     the colour as the mean <i>r</i>-fold change of every constituting transcript.
     Depicted are sign. genesets of 
     <a href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=H" target="_blank"> GSEA hallmark</a>.</p> 
     <br>

  <img src="data/Legend.png" width="100%;">

  <h2><b>How to navigate through the treemap</b></h2>

  <p>A double leftclick opens, a double rightclick closes a group. Use the mouse wheel to zoom in and out.
     Mouse over a polygon to see the full gene name and clustering hierarchy.
     Leftclick on a gene tile updates the details panel.<br>
     CTRL + leftclick on a gene tile retrieves information on the respective gene in the Ensembl database, provided pop-ups are enabled in the browser.<br>
     SHIFT + leftclick opens the UniProt database of the assigned protein.<br>
     ALT + leftclick creates an image of the displayed area.
  </p>

</div>


<!-- Boxes and text for the details panel -->

<div id="details">

  <h2>
    <span id="label"></span>
    <span> ( </span>
    <span id="Gene_Name"></span>
    <span> )</span>
  </h2>

  <p>
    <span>Mean normalized count ( </span><span id="weight" style="font-weight: bold"></span>
    <span> ) and ratio ( </span>
    <span id="ratio" style="font-weight: bold"></span>
    <span> ) and p-value ( </span>
    <span id="pvalue" style="font-weight: bold"></span>
    <span> )</span>
  </p>


<!-- BarPlot with Chart.js (which uses pixel and do not work with viewport) => <div id="chart-container" -->
<!-- y-axis text "Normalized Count" is created in html to make it responsive (in viewport) => <div id="chart-yAxis" -->
<!-- Legend (Placebo and Treated) is created in html to make it responsive (in viewport) => <div id="chart-legend" -->
  <div id="chart-box" style="width: 28vw; height: 6vw;">
    <div id="chart-yAxis" style="float: left; width: 2vw; height: 6vw; padding-top: 3.4vw;">
      <h4 style="transform: rotate(270deg)">Normalized&nbsp;Counts</h4>
    </div>

    <div id="chart-container" style="float: left; width: 20vw; height: 6vw;">
      <canvas id="detailsChart" style="width: 100%; height: 100%;"></canvas>
    </div>

    <div id="chart-legend" style="float: left; padding-left: 1vw; width: 6vw; height: 6vw;">
      <p style="color:rgb(30, 144, 255); font-size: 1.1vw"><b>Inner</b></p>
      <p style="color:rgb(238, 44, 44); font-size: 1.1vw"><b>Outer</b></p>
    </div>
  </div>



  <h2>Papio anubis</h2>
  <p><span>Ensembl ID: </span><span id="Pan_Ensembl"></span>
  <p><span>UniProt ID: </span><span id="Pan_UniProt"></span>
  <p><span>Function: </span><span id="Pan_Function"></span></p>

  <h2>Orthologous gene in Homo sapiens </h2>
  <p><span id="identical_PanToHsa" style="font-weight: bold"></span>
     <span> % identical (Papio anubis to Homo sapiens by </span>
     <span id="homology"></span>
     <span> )</span>
  <p><span id="identical_HsaToPan" style="font-weight: bold"></span>
     <span> % identical (Homo sapiens to Papio anubis) </span>     
  <p><span>Ensembl ID: </span><span id="Hsa_Ensembl"></span>
  <p><span>UniProt ID: </span><span id="Hsa_UniProt"></span>
  <p><span>Function: </span><span id="Hsa_Function"></span></p>

</div>


<script>
//Start of Foamtree-Script
  window.addEventListener("load", function () {


//Initialize the UI, input the data path here (e.g. data/Template_data.js)
    initDatasets([
      { url: 'data/Inner_Outer_Glomeruli.js', label: "" },
    ]);
    initAutocomplete();


//global parameters used by search-tool, details panel and BarPlot (must be identical with headings in data (upper and lower case do matter)
    function showGroupDetails(group) {
      var details1 = group.label;
      var details2 = group.Gene_Name;
      var details3 = group.Pan_Ensembl;
      var details4 = group.Pan_UniProt;
      var details5 = group.Pan_Function;
      var details6 = group.Hsa_Ensembl;
      var details7 = group.Hsa_UniProt;
      var details8 = group.Hsa_Function;
      var details9 = group.homology;
      var details10 = group.identical_PanToHsa;
      var details11 = group.identical_HsaToPan;
      var details12 = group.pvalue;
      var details13 = group.weight;
      var details14 = group.ratio;
      var S1 = group.Sample1;
      var S2 = group.Sample2;
      var S3 = group.Sample3;
      var S4 = group.Sample4;
      var S5 = group.Sample5;
      var S6 = group.Sample6;
      var S7 = group.Sample7;
      var S8 = group.Sample8;
      var S9 = group.Sample9;
      var S10 = group.Sample10;
        if (group.label) {
          $("#label").text(details1);
          $("#Gene_Name").text(details2);
          $("#Pan_Ensembl").text(details3);
          $("#Pan_UniProt").text(details4);
          $("#Pan_Function").text(details5);
          $("#Hsa_Ensembl").text(details6);
          $("#Hsa_UniProt").text(details7);
          $("#Hsa_Function").text(details8);
          $("#homology").text(details9);
          $("#identical_PanToHsa").text(details10);
          $("#identical_HsaToPan").text(details11);
          $("#pvalue").text(details12);
          $("#weight").text(details13);
          $("#ratio").text(details14);
          $("#Sample1").text(S1);
          $("#Sample2").text(S2);
          $("#Sample3").text(S3);
          $("#Sample4").text(S4);
          $("#Sample5").text(S5);
          $("#Sample6").text(S6);
          $("#Sample7").text(S7);
          $("#Sample8").text(S8);
          $("#Sample9").text(S9);
          $("#Sample10").text(S10);


  //BarPlot (destroy old BarPlot before new one or erros occur on hover)
            var ctx = document.getElementById('detailsChart').getContext('2d');
            if(window.bar != undefined) 
            window.bar.destroy(); 
            ctx.canvas.width = $('#graph').width();   // resize to parent width
            ctx.canvas.height = $('#graph').height(); // resize to parent height
            window.bar = new Chart(ctx , {

  //The type of chart we want to create
              type: 'bar',

  //The data for our dataset (labels for names and data like the defined variables above, e.g. S1, S2, S3, S4, S5, S6)
              data: {
                labels: ['S1', 'S2', 'S3', 'S4', 'S5', 'S6', 'S7', 'S8', 'S9', 'S10'],
                datasets: [{
                  label: 'none',
                  backgroundColor: ["rgb(30, 144, 255)","rgb(30, 144, 255)","rgb(30, 144, 255)","rgb(30, 144, 255)","rgb(30, 144, 255)","rgb(238, 44, 44)","rgb(238, 44, 44)","rgb(238, 44, 44)","rgb(238, 44, 44)","rgb(238, 44, 44)"],
                  borderColor: 'rgb(255, 255, 255)',
                  data: [S1, S2, S3, S4, S5, S6, S7, S8, S9, S10]
                }]
              },

  //Configuration options go here
              options: {
                scales: {
                  yAxes: [{
                    ticks: {
                      fontColor: 'rgb(155, 155, 155)',
                      beginAtZero: true,
                      userCallback: function(value, index, values) {
                        value = value.toString();
                        value = value.split(/(?=(?:...)*$)/);
                        value = value.join(',');
                        return value;
                      }
                    }
                  }],

                  xAxes: [{
                    ticks: {
                      fontColor: 'rgb(155, 155, 155)',
                      fontStyle: 'bold',
                    }
                  }]
                },

                legend: {display: false},
                tooltips: {
                  callbacks: {
                    label: function(tooltipItem, data) {
                      var value = data.datasets[0].data[tooltipItem.index];
                      if(parseInt(value) >= 1000){
                        return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
                      } else {
                        return value;
                      }
                    }
                  }
                }

              }
            });
  //End of BarPlot

      }
    }
//End of global parameters




//start foamtree visualization, change parameters here to change the Treemap
    var foamtree = new CarrotSearchFoamTree({
      id: "left",
      rolloutDuration: 0,
      pullbackDuration: 0,
      fadeDuration: 0,

  //"squarified" is much faster, "relaxed" looks better
      layout: "relaxed",

  //not all groups can be opened, so define a minimal size
      groupMinDiameter: 0,

  //do not show group with weight: 0
      showZeroWeightGroups: false,

  //the size of polygons is not comparable to their weight, so calculate resizing (more exact if higher but more time intensive)
      groupResizingBudget: 10000,

  //layoutByWeightOrder creates polygons in order of their weight (highest first, which looks like "fisheye")
      layoutByWeightOrder: false,

  //relaxationInitializer: "random" can be used to change visualization, "fisheye" and "blackhole" are other options

  //make border smaller
      groupStrokeWidth: 0,
      groupBorderRadius: 0,
      groupBorderWidth: 0.5,
      groupBorderWidthScaling: 1,
      groupInsetWidth: 0.5,
      groupSelectionOutlineWidth: 0.5,

  //fill color without light effects
      groupFillType: "plain",

  //no transparancy
      parentFillOpacity: 1,
      parentLabelOpacity: 1,

  //smaller polygon-text
      groupLabelMinFontSize: 1,
      groupLabelLineHeight: 1,

  //left and right space at name edges
      groupLabelHorizontalPadding: 0.5,
      groupLabelVerticalPadding: 0.5,

  //draw all levels
      maxGroupLevelsDrawn: 4,
      maxGroupLabelLevelsDrawn: 4,

  //ZoomFaktor
      zoomMouseWheelFactor: 1.3,

  //Zoom instead of animation to improve performance once loaded
      onModelChanging: function addParent(group, parent) {
        group.parent = parent;
        if (group.groups) {
          group.groups.forEach(function(g) {
            addParent(g, group);
          });
        }
      },

      onGroupDoubleClick: function (e) {
        e.preventDefault();
        var group = e.secondary ? e.bottommostOpenGroup : e.topmostClosedGroup;
        var toZoom;
        if (group) {
          this.open({ groups: group, open: !e.secondary });
        } else {
        }
      },

  //no colorchange if hovered
      groupHoverFillLightnessShift: 0,

  //text if hovered
      titleBar: "inscribed",
        maxLabelSizeForTitleBar: Number.MAX_VALUE,
        titleBarDecorator: function(options, parameters, variables) {
          var Hierarchy = parameters.group.Hierarchy;
          if (Hierarchy === 1) {
            variables.titleBarText = parameters.group.label
          } else {
            if (Hierarchy === 2) {
              variables.titleBarText = parameters.parent.label + " > " + parameters.group.label
            } else {
              if (Hierarchy === 3) {
                variables.titleBarText = parameters.parent.parent.label + " > " + parameters.parent.label + " > " + parameters.group.label
              } else {
                variables.titleBarText = parameters.parent.parent.parent.label + " > " + parameters.parent.parent.label + " > " + parameters.parent.label + " > " + parameters.group.label + " > " + parameters.group.Gene_Name
              }
            }
          }
        },

  //Alt+Click for an image, CTRL+Click and SHIFT+Click for for Hyperlink to Ensembl and UniProt database
      onGroupClick: function (event) {
        var group = event.group;
        showGroupDetails(group);

          if (event.ctrlKey) {
            event.preventDefault();
            open("http://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=" + group.Hsa_Ensembl)
          }

          if (event.shiftKey) {
            event.preventDefault();
            open("https://www.uniprot.org/uniprot/" + group.Hsa_UniProt)
          }

          if (event.altKey) {
            event.preventDefault();
            window.open(foamtree.get("imageData", {
            format: "image/png", quality: 1, pixelRatio: 10 }));
          }
      },

  //manual color coding of each polgon
      groupColorDecorator: function (opts, params, vars) {
        vars.groupColor = params.group.color;
        vars.labelColor = "auto";
      },

  //This listener processes the input data before display.
      onModelChanging: function (dataObject) {
        CarrotSearchFoamTree.TreeModel.eachDescendantAndSelf(dataObject, function (group, parent, level) {

  //Remove the loading-box when foamtree is ready
          $(".loading").remove();

  //If no label, use the id as the label
          if (!group.label) {
            group.label = group.id + "";
          }
        });
      },

//End of "var foamtree = new CarrotSearchFoamTree({"
    });




//Script for all features, no change needed
//Zoom to the group when form submitted (label as search criteria)
    $("#form").submit(function (e) {
      e.preventDefault();
      var query = $("#search").val();

      if ($.trim(query).length > 0) {
        var group = CarrotSearchFoamTree.TreeModel.findFirstByProperty(foamtree.get("dataObject"), "label", query);

        if (group) {
          zoomToGroup(group);
          showGroupDetails(group);
        }
      } else {
        zoomToGroup(foamtree.get("dataObject"));
      }
    });

//Zoom to the group when form submitted (Gene_Name as search criteria)
    $("#form").submit(function (e) {
      e.preventDefault();
      var query = $("#search").val();

      if ($.trim(query).length > 0) {
        var group = CarrotSearchFoamTree.TreeModel.findFirstByProperty(foamtree.get("dataObject"), "Gene_Name", query);

        if (group) {
          zoomToGroup(group);
          showGroupDetails(group);
        }
      } else {
        zoomToGroup(foamtree.get("dataObject"));
      }
    });

//Handle clicks on search autocomplete
    $("body")
      .on("click", "a.dataset",function (e) {
        e.preventDefault();
        loadDataSet(this.href);
      })
      .on("click", "a[href = '#zoom-out']",function (e) {
        e.preventDefault();
        zoomToGroup(foamtree.get("dataObject"));
      })
      .on("click", "a.go", function (e) {
        e.preventDefault();
        var id = this.hash.substring(1);
        var group = CarrotSearchFoamTree.TreeModel.findFirstByProperty(foamtree.get("dataObject"), "id", id);
        if (group) {
          zoomToGroup(group);
        }
      });

//Load the first data set
    $("a.dataset:eq(0)").trigger("click");



//
//Utility functions
//
    function initDatasets(datasets) {
      var template = Template.make("<a href='<%- url %>' class='dataset'><%- label %></a>");

      $("#datasets").html(datasets.reduce(function (html, ds) {
        html += template(ds);
        return html;
      }, ""));
    }

//Initiate loading of the data
    function loadDataSet(dataSetUrl) {
      var $loading = $("#loading").show();
      var $ui = $("#ui").hide();
      JSONP.load(dataSetUrl, "modelDataAvailable", function (dataObject) {
        prepareDirectLinks(dataObject);

        $loading.hide();
        $ui.show();
        lastZoomedTo = undefined;

        setTimeout(function () {
          foamtree.set("dataObject", dataObject);
        }, 100);
      });
    }

    function prepareDirectLinks(dataObject, byName) {
      var template = Template.make("<a class='go' href='#<%- id %>'><%- label %></a>");
      var templateWithLevel = Template.make("<a class='go' href='#<%- id %>'><%- label %> (level:&nbsp;<%- level %>)</a>");
      var templateWithChildren = Template.make("<a class='go' href='#<%- id %>'><%- label %> (children:&nbsp;<%- children %>)</a>");

      var count = 0, numLeafGroups = 0, numNonLeafGroups = 0, maxChildren = 0;
      CarrotSearchFoamTree.TreeModel.eachDescendantAndSelf(dataObject, function (group, index, parent, level) {
        if (group.id === undefined) {
          group.id = count;
        }
        group.parent = parent;
        group.level = level;
        group.children = group.groups ? group.groups.length : 0;

        if (maxChildren < group.children) {
          maxChildren = group.children;
        }
        count++;
        if (group.children > 0) {
          numNonLeafGroups++;
        } else {
          numLeafGroups++;
        }
      });

      var html = "";

  //Deepest nesting level
      var maxLevel = 0, maxLevelGroups = [];
      CarrotSearchFoamTree.TreeModel.eachDescendantAndSelf(dataObject, function (group, index, parent, level) {
        if (maxLevel < level) {
          maxLevel = level;
        }
      });
      var maxLevelTmp = maxLevel;
      for (var i = 0; i < 4; i++) {
        CarrotSearchFoamTree.TreeModel.eachDescendantAndSelf(dataObject, function (group, index, parent, level) {
          if (group.label && level === maxLevelTmp) {
            maxLevelGroups.push(group);
            return false;
          }
        });
        maxLevelTmp -= Math.ceil(maxLevelTmp / 20);
      }

      html += "<div>Deeply nested groups: " + maxLevelGroups.reduce(function (html, group) {
        html += templateWithLevel(group);
        return html;
      }, "") + "</div>";

  //Many children
      var maxChildrenGroups = [];
      CarrotSearchFoamTree.TreeModel.eachDescendantAndSelf(dataObject, function (group, index, parent, level) {
        if (group.label) {
          maxChildrenGroups.push(group);
        }
      });
      maxChildrenGroups.sort(function (a, b) { return b.children - a.children });

      html += "<div>Groups with many children: " + maxChildrenGroups.slice(0, 5).reduce(function (html, group) {
        html += templateWithChildren(group);
        return html;
      }, "") + "</div>";

      $("#direct").html(html);

      var statsTemplate = Template.make("<dl class='dl-horizontal dl-wide'>" +
        "<dt>Total number groups:</dt><dd><%- numGroups %></dd>" +
        "<dt>Non-leaf groups:</dt><dd><%- numNonLeafGroups %></dd>" +
        "<dt>Leaf groups:</dt><dd><%- numLeafGroups %></dd>" +
        "<dt>Deepest nesting level:</dt><dd><%- maxLevel %></dd>" +
        "<dt>Max children in one group:</dt><dd><%- maxChildren %></dd>" +
      "</dl>");

      $("#stats").html(statsTemplate({
        numGroups: count,
        numNonLeafGroups: numNonLeafGroups,
        numLeafGroups: numLeafGroups,
        maxLevel: maxLevel,
        maxChildren: maxChildren
      }));

//End of "function prepareDirectLinks(dataObject, byName) {"
    }


//Zoom function
    function zoomToGroup(target) {
      if (target === lastZoomedTo) {
        return;
      }

      if (!lastZoomedTo) {
        lastZoomedTo = foamtree.get("dataObject");
      }

  //Find the lowest parent ancestor between target and lastZoomedTo
      var common = lowestCommonAncestor(lastZoomedTo, target);

  //We'll need to open all parents of the node we're zooming to. Very deeply-nested nodes will only be initialized if their parents get open. If we don't open the parents, the polygon corresponding to the target group would not exist.
      var targetParents = allParents(target);
      targetParents.pop();

  //It may happen that our target site is so deeply-nested, FoamTree could not compute a diagram for it due to insufficient numeric precision.
  //In this case, we'll zoom to the closest computed parent. The label decorator will draw a [+] annotation to indicate that the site has children whose diagrams could not be computed.
      foamtree.open(targetParents).then(function() {
        while (target.parent && !foamtree.get("state", target.parent).browseable) {
          target = target.parent;
        }

    //First, zoom out to the lowest common ancestor. Temporarily set the zoom duration based on the difference in levels we need to travel.
        foamtree.set("zoomMouseWheelEasing", "squareInOut");
        foamtree.set("zoomMouseWheelDuration", Math.max(2000, (lastZoomedTo.level - common.level) * 50));
        foamtree.zoom(common).then(function() {

      //Open all parents of the target group.
          foamtree.open({ groups: target, open: false });
          foamtree.set("zoomMouseWheelDuration", Math.max(2000, (target.level - common.level) * 100));

      //Zoom to the target group.
          foamtree.zoom(target);
          foamtree.set("zoomMouseWheelDuration", CarrotSearchFoamTree.defaults.zoomMouseWheelDuration);
          foamtree.set("zoomMouseWheelEasing", CarrotSearchFoamTree.defaults.zoomMouseWheelEasing);
        });
        lastZoomedTo = target;
      });

//End of "function zoomToGroup(target) {"
    }


//Search function
    function lowestCommonAncestor(groupA, groupB) {
      var parentsA = allParents(groupA);
      var parentsB = allParents(groupB);

      var max = Math.min(parentsA.length, parentsB.length);
      for (var i = 0; i < max; i++) {
        if (parentsA[i] !== parentsB[i]) {

  //We assume the two nodes do have one common parent
          return parentsA[i - 1];
        }
      }
      return parentsA[max - 1];
    }

//Search function
    function allParents(group) {
      var parents = [];
      var parent = group;
      while (parent) {
        parents.push(parent);
        parent = parent.parent;
      }
      parents.reverse();
      return parents;
    }

//Suggestions for Search_Autocomplete
    function initAutocomplete() {
      var templates = {
        suggestion: Template.make("<div><%- group.label %><small><%- group.Geneset %></small><small><%- group.Gene_Name %></small></div>")
      };

      $('#search').typeahead({
        hint: true,
        highlight: true,
        minLength: 1
      },
      {
        name: 'labels',
        displayKey: 'value',
        source: foamTreeSource("label"),
        templates: templates
      },
      {
        name: 'Geneset',
        displayKey: 'value',
        source: foamTreeSource("Geneset"),
        templates: templates
      },
      {
        name: 'Gene_Name',
        displayKey: 'value',
        source: foamTreeSource("Gene_Name"),
        templates: templates
      }
      );

  //Queries FoamTree model for the purposes of the autocomplete input.
      function foamTreeSource(prop) {
        return function findMatches(q, cb) {
          var matches = [];
          q = q.toLowerCase();

          CarrotSearchFoamTree.TreeModel.eachDescendantAndSelf(foamtree.get("dataObject"), function (group) {
            var val = (group[prop] + "");
            var index = val.toLocaleLowerCase().indexOf(q);
            if (index >= 0) {
              matches.push({ value: val, index: index, group: group });
            }
          });

          matches.sort(function (a, b) {
            if (a.index != b.index) {
              return a.index - b.index;
            } else {
              if (a.group.label < b.group.label) {
                return -1;
              } else if (a.group.label > b.group.label) {
                return 1;
              } else {
                return 0;
              }
            }
          });

      //circa the amount of Search_Suggestions
          cb(matches.slice(0, 20));
        };
      }

//End of "function initAutocomplete() {"
    }

//End of "window.addEventListener("load", function () {"
  });



//Show/Hide details panel

  var detailsshown = true;

  function showdetails() {
    var btn = document.getElementById("detailsbutton");
    if(detailsshown === true){
      detailsshown = false;
      btn.value = 'A'; btn.innerHTML = '<<< Details';

      var elem = document.getElementById("details");   
      var pos = 0;
      var id = setInterval(frame, 20);
      function frame() {
        if (pos == 30) {
          clearInterval(id);
        } else {
          pos++; 
          elem.style.left = 37.5 + pos + "vw"; 
        }
      }

    }else{
      detailsshown = true;
      btn.value = 'A'; btn.innerHTML = 'Details >>>';

      var elem = document.getElementById("details");   
      var pos = 0;
      var id = setInterval(frame, 20);
      function frame() {
        if (pos == 30) {
          clearInterval(id);
        } else {
          pos++; 
          elem.style.left = 67.5 - pos + "vw"; 
        }
      }
    }
  }

</script>

</body>
</html>